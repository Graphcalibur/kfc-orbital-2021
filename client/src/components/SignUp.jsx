import React, { Component } from "react";
import { Button, Form, Modal } from "react-bootstrap";

class SignUp extends Component {
  state = {
    username: "",
    email: "",
    password: "",
    confirm_password: "",
    validated: false,
    failed_sign_up: false,
  };

  handleUsernameChange = (event) => {
    this.setState({ username: event.target.value });
  };

  handleEmailChange = (event) => {
    this.setState({ email: event.target.value });
  };

  handlePasswordChange = (event) => {
    this.setState({ password: event.target.value });
  };

  handleConfirmChange = (event) => {
    this.setState({ confirm_password: event.target.value });
  };

  /* Checks if all input fields in the form are valid and if so,
  attempts to register the user. */
  handleSubmit = (event) => {
    const form = event.currentTarget;
    event.preventDefault();
    event.stopPropagation();
    this.setState({ validated: true });

    if (
      !form.checkValidity() ||
      this.state.password !== this.state.confirm_password
    )
      return;

    this.registerUser();
  };

  /* Registers the user. If sign up was successful, then it automatically
  logins the user and refreshes the page. Otherwise, it triggers the
  failed sign up message. */
  registerUser = () => {
    const requestOptions = {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
        username: this.state.username,
        password: this.state.password,
      }),
    };

    const requestOptions2 = {
      ...requestOptions,
      credentials: "include",
    };

    /* Automatically logs the user in after successful sign up. If sign up is
      not successful, displays an error message */
    fetch("/api/register", requestOptions).then((res) => {
      if (res.status !== 409) {
        fetch("/api/authuser", requestOptions2).then((res) => {
          window.location.reload();
        });
      } else {
        this.setState({ failed_sign_up: true });
      }
    });
  };

  failedSignUpText = () => {
    return this.state.failed_sign_up ? (
      <p style={{ fontSize: "80%", color: "#ff1a1a" }}>
        Sign up has failed. Someone with that username already exists. Please
        choose a different username.
      </p>
    ) : (
      <></>
    );
  };

  render() {
    const { password, confirm_password } = this.state;
    return (
      <Modal
        show={this.props.show}
        onHide={this.props.close}
        backdrop="static"
        keyboard={false}
        className="modal"
      >
        <Modal.Header closeButton>
          <Modal.Title>
            {this.state.registered ? "You have signed up!" : "Sign Up"}
          </Modal.Title>
        </Modal.Header>

        <Modal.Body>
          <Form
            noValidate
            validated={this.state.validated}
            onSubmit={this.handleSubmit}
          >
            <Form.Group className="mb-3" controlId="username">
              <Form.Label>Username</Form.Label>
              <Form.Control
                required
                placeholder="Username"
                value={this.state.username}
                onChange={(event) => this.handleUsernameChange(event)}
              />
              <Form.Control.Feedback type="invalid">
                Please enter a valid username.
              </Form.Control.Feedback>
            </Form.Group>

            <Form.Group className="mb-3" controlId="email">
              <Form.Label>Email (Optional)</Form.Label>
              <Form.Control
                type="email"
                placeholder="Email"
                value={this.state.email}
                onChange={(event) => this.handleEmailChange(event)}
              />
              <Form.Control.Feedback type="invalid">
                Please enter a valid email.
              </Form.Control.Feedback>
            </Form.Group>

            <Form.Group className="mb-3" controlId="password">
              <Form.Label>Password</Form.Label>
              <Form.Control
                required
                type="password"
                placeholder="Password"
                value={password}
                onChange={(event) => this.handlePasswordChange(event)}
              />
              <Form.Control.Feedback type="invalid">
                Please enter a valid password.
              </Form.Control.Feedback>
            </Form.Group>

            <Form.Group className="mb-3" controlId="confirm" id="confirm">
              <Form.Label>Confirm Password</Form.Label>
              <Form.Control
                required
                type="password"
                placeholder="Confirm Password"
                value={confirm_password}
                onChange={(event) => this.handleConfirmChange(event)}
                isInvalid={password !== confirm_password}
              />
              <Form.Control.Feedback type="invalid">
                This must be the same as your password.
              </Form.Control.Feedback>
            </Form.Group>

            {this.failedSignUpText()}

            <Button variant="primary" type="submit" className="float-end">
              Sign Up
            </Button>
          </Form>
        </Modal.Body>
      </Modal>
    );
  }
}

export default SignUp;
