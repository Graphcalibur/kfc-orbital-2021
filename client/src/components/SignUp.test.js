import React from "react";
import SignUp from "./SignUp";
import { create } from "react-test-renderer";

describe("Testing SignUp component", () => {
  test("Test 1: Not shown", () => {
    const tree = create(<SignUp show={false} close={() => null} />);
    expect(tree).toMatchSnapshot();
  });

  test("Test 2: Test validation with empty at least one empty field", () => {
    const tree = create(<SignUp show={true} close={() => null} />);

    /* Test username filled, rest empty */
    const instance = tree.getInstance();
    instance.setState({ username: "test", validated: true });
    expect(tree.toJSON()).toMatchSnapshot();

    /* Test password and confirm_password filled, rest empty */
    instance.setState({
      username: "",
      password: "test",
      confirm_password: "test",
      validated: true,
    });
    expect(tree.toJSON()).toMatchSnapshot();
  });

  test("Test 3: Test validation with filled in fields but different password and confirm_password", () => {
    const tree = create(<SignUp show={true} />);
    const instance = tree.getInstance();
    instance.setState({
      username: "test",
      password: "testtest",
      confirm_password: "doy",
      validated: true,
    });
    expect(tree.toJSON()).toMatchSnapshot();
  });

  test("Test 4:  Test validation with all fields valid except for email", () => {
    const tree = create(<SignUp show={true} />);
    const instance = tree.getInstance();
    instance.setState({
      username: "test",
      email: "lol",
      password: "testtest",
      confirm_password: "testtest",
      validated: true,
    });
    expect(tree.toJSON()).toMatchSnapshot();
  });

  test("Test 5:  Test validation with all fields valid, email empty", () => {
    const tree = create(<SignUp show={true} />);
    const instance = tree.getInstance();
    instance.setState({
      username: "test",
      password: "testtest",
      confirm_password: "testtest",
      validated: true,
    });
    expect(tree.toJSON()).toMatchSnapshot();
  });

  test("Test 6:  Test validation with all fields valid and email filled in", () => {
    const tree = create(<SignUp show={true} />);
    const instance = tree.getInstance();
    instance.setState({
      username: "test",
      email: "lol@gmail.com",
      password: "testtest",
      confirm_password: "testtest",
      validated: true,
    });
    expect(tree.toJSON()).toMatchSnapshot();
  });

  test("Test 7:  Test validation with failed sign up", () => {
    const tree = create(<SignUp show={true} />);
    const instance = tree.getInstance();
    instance.setState({
      username: "test",
      email: "lol@gmail.com",
      password: "testtest",
      confirm_password: "testtest",
      validated: true,
      failed_sign_up: 1,
    });
    expect(tree.toJSON()).toMatchSnapshot();
  });
});
