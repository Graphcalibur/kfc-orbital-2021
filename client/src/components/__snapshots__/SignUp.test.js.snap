// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing SignUp component Test 1: Not shown 1`] = `null`;

exports[`Testing SignUp component Test 2: Test validation with empty at least one empty field 1`] = `null`;

exports[`Testing SignUp component Test 2: Test validation with empty at least one empty field 2`] = `null`;

exports[`Testing SignUp component Test 3: Test validation with filled in fields but different password and confirm_password 1`] = `null`;

exports[`Testing SignUp component Test 4:  Test validation with all fields valid except for email 1`] = `null`;

exports[`Testing SignUp component Test 5:  Test validation with all fields valid, email empty 1`] = `null`;

exports[`Testing SignUp component Test 6:  Test validation with all fields valid and email filled in 1`] = `null`;

exports[`Testing SignUp component Test 7:  Test validation with failed sign up 1`] = `null`;
